src/main/java/org/apache/aurora/scheduler/updater/strategy/VariableBatchStrategy.java

```diff
@@ -0,0 +1,152 @@
+
+    LOG.debug("Variable Batch Update progress: {} instances have been modified, "
+            + "{} instances remain unmodified, and {} overall instances to be modified.",
+        modified,
+        remaining,
+        totalModInstanceCount.get());
+
+    if (rollingForward) {
+
+      int sum = 0;
+      for (Integer groupSize : maxActiveGroups) {
+        sum += groupSize;
+
+        if (sum > modified) {
+          return groupSize;
+        }
+      }
+      // Return last step when number of instances > sum of all groups
+      return lastGroupSize;
+    } else {
+
+      // To perform the update in reverse, we use the number of remaining tasks left to update
+      // instead of using the number of already modified instances. In a rollback, the remaining
+      // count represents the number of instances that were already modified while rolling forward
+      // and need to be reverted.
+      int curGroupSize = remaining;
+
+      for (Integer groupSize : maxActiveGroups) {
+        // This handles an in between step. i.e.: updated instances = 4, update groups = [2,3]
+        // which results in update groups 2 and 2 rolling forward at the time of failure.
+        if (curGroupSize <= groupSize) {
+          return curGroupSize;
+        }
+
+        curGroupSize -= groupSize;
+      }
+
+      // Handle the case where number of instances update were
+      // greater than the sum of all update groups
+      // Calculate the size of the last update group size performed while rolling forward.
+      curGroupSize = curGroupSize % lastGroupSize;
+      if (curGroupSize == 0) {
+        return lastGroupSize;
+      } else {
+        return curGroupSize;
+      }
+    }
+  }
+
+  @Override
+  public final Set<T> getNextGroup(Set<T> idle, Set<T> active) {
+
+    // Get the size for the idle set on the first run only. This is representative of the number
+    // of overall instance modifications this update will trigger.
+    if (!totalModInstanceCount.isPresent()) {
+      totalModInstanceCount = Optional.of(idle.size());
+    }
+
+    // Limit group size to the current size of the group minus the number of instances currently
+    // being modified.
+    return ordering.sortedCopy(doGetNextGroup(idle, active)).stream()
+            .limit(Math.max(0, determineCurGroupSize(idle.size()) - active.size()))
+            .collect(Collectors.toSet());
+  }
+
+  /**
+   * Return a list of instances to be updated.
+   * Returns an empty list if the current active group has not completed.
+   * If the result is larger than the current group size in {@link #maxActiveGroups},
+   * it will be truncated.
+   *
+   * @param idle Idle instances, candidate for being updated.
+   * @param active Instances currently being updated.
+   * @return A subset of {@code idle}, instances to start updating.
+   */
+  Set<T> doGetNextGroup(Set<T> idle, Set<T> active) {
+    if (active.isEmpty()) {
```

Comment:
I would copy the current `BatchStrategy` impl of this since it is the same:
`return active.isEmpty() ? idle : ImmutableSet.of();`